fvm_actor_utils/src/receiver/mod.rs

```diff
@@ -0,0 +1,229 @@
+use std::mem;
+
+use fvm_ipld_encoding::tuple::{Deserialize_tuple, Serialize_tuple};
+use fvm_ipld_encoding::{Cbor, RawBytes};
+use fvm_shared::{address::Address, econ::TokenAmount, error::ExitCode};
+use num_traits::Zero;
+use thiserror::Error;
+
+use crate::messaging::{Messaging, MessagingError, RECEIVER_HOOK_METHOD_NUM};
+use crate::receiver::frc46::{FRC46TokenReceived, FRC46_TOKEN_TYPE};
+
+pub mod frc46;
+
+/// Standard interface for an actor that wishes to receive FRC-0046 tokens or other assets
+pub trait UniversalReceiver {
+    /// Invoked by a token actor during pending transfer or mint to the receiver's address
+    ///
+    /// Within this hook, the token actor has optimistically persisted the new balance so
+    /// the receiving actor can immediately utilise the received funds. If the receiver wishes to
+    /// reject the incoming transfer, this function should abort which will cause the token actor
+    /// to rollback the transaction.
+    fn receive(params: UniversalReceiverParams);
+}
+
+/// Type of asset received - could be tokens (FRC46 or other) or other assets
+pub type ReceiverType = u32;
+
+#[derive(Error, Debug)]
+pub enum ReceiverHookError {
+    #[error("receiver hook was not called")]
+    NotCalled,
+    #[error("receiver hook was already called")]
+    AlreadyCalled,
+    #[error("error encoding to ipld")]
+    IpldEncoding(#[from] fvm_ipld_encoding::Error),
+    #[error("error sending message")]
+    Messaging(#[from] MessagingError),
+    #[error("receiver hook error from {address:?} when called with {receiver_params:?}: exit_code={exit_code:?}, return_data={return_data:?}")]
+    Receiver {
+        address: Address,
+        exit_code: ExitCode,
+        return_data: RawBytes,
+        receiver_params: UniversalReceiverParams,
```

Comment:
This error replaces the old 
```
 ReceiverHook {
        /// Whose balance is being debited
        from: ActorID,
        /// Whose balance is being credited
        to: ActorID,
        /// Who initiated the transfer of funds
        operator: ActorID,
        amount: TokenAmount,
        exit_code: ExitCode,
    },
``` 

The new error returns more information about the message itself and the receiving actor rather than semantics related to the transfer

Reply:
When this error is returned, we should still have the original `ReceiverHook` struct available so I don't think we need the `UniversalReceiverParams`. Might be worth keeping the type from it for error messages, then adding some getter functions to `ReceiverHook` to query other data (or making some fields public maybe).

The important data here would be the `exit_code` and `return_data`, because they're not available elsewhere. Return data might not be useful either if the hook aborts, I don't think we get any return data in that case?

Reply:
Makes sense - and it seems this was the error in question that was too large so removing it also fixes the CI error